### ISSUE TITLE
Add `rel=canonical` support

### ISSUE THREAD

**Issue by briansmith** [rust-lang/docs.rs#74]
Please add the correct [`<link rel=canonical>`](https://en.wikipedia.org/wiki/Canonical_link_element) link to each generated documentation page, when the crate provides a canonical documentation URL. This will help search engines disambiguate the docs.rs copies of the documentation and avoid docs.rs looking like a content farm or spam blog to search engines.

**Comment by onur** [comment]
Thanks for this awesome idea. I didn't know about `rel=canonical`.

I'll definitely add this, but I need to save original documentation links first. That is only thing I forgot to save when building crates.

**Comment by jyn514** [comment]
@GuillaumeGomez is this something that should be done by docs.rs? It seems pretty easy to implement but I think it would make more sense to have a `rel=canonical` link in every page generated with rustdoc, not just the ones on docs.rs.

**Comment by GuillaumeGomez** [comment]
If I understand correctly, it's supposed to be used in the `<head>` part. However, I'm not sure if this'll be really useful here considering that all URLs are unique in our case... Also, the URL needs to include the domain name, so if we want to add it, I think it should be done on docs.rs side.

**Comment by jyn514** [comment]
This is (correct me if I'm wrong) for documentation URLs that are explicitly given in Cargo.toml, not for different versions of the docs. See for example https://github.com/serde-rs/serde/blob/master/serde/Cargo.toml#L9. The domain name is given here, it's not necessarily the same site that's currently hosting the docs.

**Comment by GuillaumeGomez** [comment]
Oh in this case it makes sense to have canonical then I guess. But still on docs.rs side. :)

**Comment by workingjubilee** [comment]
P-higher than medium maybe but not existentially critical so not high ( I mean I guess that's the important label? ), `https://docs.rs/tracing/0.1.1.9/tracing` existing punishes e.g. `https://tracing.rs/tracing`. This needs to be resolved sooner rather than later.

**Comment by workingjubilee** [comment]
I don't know for certain that for an explicitly specified documentation field the `<link rel="canonical">` **shouldn't** be fixed on the rustdoc side given that the most natural place to do this is at build time, since it is extracted from the Cargo.toml and resolves to a URL. Because it is, it is actually [part of crates.io's public API](https://doc.rust-lang.org/cargo/reference/manifest.html#the-documentation-field) to link to that URL.

**Comment by jyn514** [comment]
@workingjubilee this is part of cargo.toml, not the source code. It's possible to add an unstable rustdoc flag but since we'd be the only ones using it I'd rather just inject it with the rest of the styles.

**Comment by workingjubilee** [comment]
Whoops, yeah, that would have to be as part of cargo doc and no deeper, ideally.

**Comment by pietroalbini** [comment]
Hmm, I'm not sure relying on the `package.documentation` key of `Cargo.toml` is the best idea for this: that key could point to (for example) a mdbook, and will probably break when adding `rel="canonical"` to subpages.

For people that explicitly want this I'd add something like this:

```toml
[package.metadata.docs.rs]
canonical-url = "https://api.example.com/{version}/{path}"
```

**Comment by pietroalbini** [comment]
An explicit metadata key with placeholders could also be used to display a banner in the UI:

![2020-08-28--10-58-56](https://user-images.githubusercontent.com/2299951/91542058-8dc45880-e91d-11ea-997a-5fab02bbe404.png)

**Comment by jyn514** [comment]
> that key could point to (for example) a mdbook, and will probably break when adding rel="canonical" to subpages.

I was imagining it would point to the landing page even from sub pages. That way you wouldn't get spurious 404s.

I like the idea of a banner with a link that opens in a new tab though, that seems much more discoverable.

@briansmith do you have a preference of what this should look like?

**Comment by pietroalbini** [comment]
> I was imagining it would point to the landing page even from sub pages. That way you wouldn't get spurious 404s.

That's not how `rel="canonical"` works: a page with that tag should have the same contents as the canonical one. We can assume `https://docs.rs/rustwide/struct.Workspace.html` and `https://example.com/rustwide/struct.Workspace.html` have roughly the same contents, but `https://docs.rs/rustwide/struct.Workspace.html` and `https://example.com/rustwide/` definitely do not have the same content in them.

---

I thought about this a bit more, and here's is how I would implement this.

### Goals

* We should respect the crate author's wishes to have their documentation hosted outside of docs.rs.
* We should ensure people can always view the documentation on docs.rs for every version if they explicitly wish to.

### `rel="canonical"`

If the crate doesn't specify a `package.metadata.docs.rs.canonical-url`, the canonical URL will always be `https://docs.rs/{crate}/latest/{path}`. When an user visits that page, docs.rs will not redirect, but instead serve the latest version of that page.

If a `package.metadata.docs.rs.canonical-url` is specified, the canonical URL will be that path, with `{path}` replaced with the file path. If `{path}` is not present the path will be appended at the end of the URL.

### Banner

If a `package.metadata.docs.rs.canonical-url` is specified, on **the latest version** we will display a banner similar to this:

![2020-08-28--11-02-33](https://user-images.githubusercontent.com/2299951/91542340-ff9ca200-e91d-11ea-8072-0f61fca9476a.png)

When the user clicks on it they will get redirected to the canonical URL. The banner will be collapsable, and that preference will be stored on the user's browser (each crate will have its own separate cookie to collapse it). Since the metadata won't allow to specify a version, the banner will only be shown on the latest version.

**Comment by jyn514** [comment]
> Since the metadata won't allow to specify a version, the banner will only be shown on the latest version.

This doesn't seem right - metadata is per-release, not per crate. So any release that specifies a canonical URL should have the banner.

> Hmm, I'm not sure relying on the package.documentation key of Cargo.toml is the best idea for this: that key could point to (for example) a mdbook, and will probably break when adding rel="canonical" to subpages.

> [The way] rel="canonical" works [is] a page with that tag should have the same contents as the canonical one.

Ok, that makes sense, but I don't want to introduce yet another docs.rs specific toggle. There are a lot of existing crates using `package.documentation` and it's standardized by Cargo, I don't want to make a new button if we can avoid it. Maybe we could only add `rel=canonical` for `canonical-url`, but still add the banner for `package.documentation`? That way we wouldn't have incorrect `rel=canonical` but it would still be clear that the author has a preferred documentation site without having to jump through hoops.

All the other suggestions about the banner seem fine to me, although I'd be interested in Brian's opinion.

**Comment by pietroalbini** [comment]
> This doesn't seem right - metadata is per-release, not per crate. So any release that specifies a canonical URL should have the banner.

That only works if the crate owner publishes each version to a different URL and updates the URL in the `Cargo.toml` every time. Neither [ring](https://briansmith.org/rustdoc/ring/) nor [tracing](https://tracing.rs), for example, do so. Assuming that someone who views the documentation for an old version is explicitly interested in that version, the banner will point to a completly different thing. As an user, I wouldn't want to get a banner to the tokio 0.2 docs if I'm visiting the 0.1 docs.

> Maybe we could only add rel=canonical for canonical-url, but still add the banner for package.documentation? That way we wouldn't have incorrect rel=canonical but it would still be clear that the author has a preferred documentation site without having to jump through hoops.

Then we need to tweak the wording of the banner to something like "the author also provides additional documentation on [domain name]". That makes sense and it would be very useful to link (for example) [to the serde book](https://serde.rs/), but it has a completly different meaning than the banner in my mockup.

**Comment by workingjubilee** [comment]
"The authors of this crate prefer to host their documentation on (website)" sounds weird and a little passive-aggressive, frankly.

And a crate author preferring to host a new version is not a big deal. Users actively complain when they find old versions. That's kind of the entire point of wanting to canonicalize anything at all: It makes Rust look horribly dated and incapable of organizing its own documentation when the first result in searches is something from 2015~2016 even though several major versions have come and gone since then. So while other issues apply, sure, there is _not_ a problem with pointing to an author's preferred up-to-date version.

**Comment by pietroalbini** [comment]
> "The authors of this crate prefer to host their documentation on (website)" sounds weird and a little passive-aggressive, frankly.

That's totally fair, I didn't spend much time thinking about the message in a mockup :)

> And a crate author preferring to host a new version is not a big deal. Users actively complain when they find old versions. That's kind of the entire point of wanting to canonicalize anything at all: It makes Rust look horribly dated and incapable of organizing its own documentation when the first result in searches is something from 2015~2016 even though several major versions have come and gone since then. So while other issues apply, sure, there is not a problem with pointing to an author's preferred up-to-date version. It's quite relevant to point to it even on older versions because that provides information on what the author actually intends to support.
>
> Many times people have asked me how to fix a problem with a crate and I have suggested, several times, of "try using a newer version" and lo and behold...! It would be slightly extreme to make the tokio 0.1 docs cease to exist so I stop finding them, yet that would much better match the typical desire.

If we just want to set `rel="canonical"` to point at docs.rs's latest version in the initial iteration that's also ok! That should solve the immediate problem.

**Comment by jyn514** [comment]
I am strongly against setting rel=canonical to docs.rs without looking at package.documentation.

> https://docs.rs/tracing/0.1.19/tracing existing punishes e.g. `https://tracing.rs/tracing`.

We should fix that before we try and improve our own SEO.

**Comment by pietroalbini** [comment]
We can't use the value of `package.documentation` as the `rel="canonical"` though. I [extracted the value of the field across the latest published versions of every crate](https://gist.github.com/pietroalbini/d9d38961edfe4f559ecc2ddd122edde8), and while some of them indeed point to the API documentation, there are also a bunch that link to non-rustdoc documentation (like an user guide) or to the GitHub repository/README.

**Comment by jsha** [comment]
As noted in #1438, lack of rel="canonical" seems to be hurting search results within docs.rs too. Given a plethora of identical content across various releases of a crate, Google chooses one effectively at random, which can cause a page from a newer version of a crate to be excluded in favor of that same page from an older version of a crate. I have a two-part proposal:

 - For crates with no package.documentation, or where package.documentation starts with `https://docs.rs/`, set `rel="canonical"` to the equivalent page on docs.rs with `/latest/` in the URL. For example https://docs.rs/regex/1.5.5/regex/struct.CaptureLocations.html would point to https://docs.rs/regex/latest/regex/struct.CaptureLocations.html.
 
  - For crates with a package.documentation that does not start with `https://docs.rs`, set no `rel="canonical"`, and set `<meta name="robots" content="noindex">` (https://developers.google.com/search/docs/advanced/crawling/block-indexing). This should prevent the docs.rs documentation from competing with self-hosted documentation for canonical status, while leaving the docs.rs pages available for users who navigate there directly.

Whichever tag we emit, we would do it in docs.rs at page load time, not in rustdoc and not at doc build time. This allows us to include the tags even for old releases, and to remove them easily if they turn out not to have the effect we want.

**Comment by jsha** [comment]
>  - For crates with a package.documentation that does not start with `https://docs.rs`, set no `rel="canonical"`, and set `<meta name="robots" content="noindex">` (https://developers.google.com/search/docs/advanced/crawling/block-indexing). This should prevent the docs.rs documentation from competing with self-hosted documentation for canonical status, while leaving the docs.rs pages available for users who navigate there directly.

An update on this: https://developers.google.com/search/docs/advanced/crawling/consolidate-duplicate-urls says:

> **Don't** use [noindex](https://developers.google.com/search/docs/advanced/crawling/block-indexing) as a means to prevent selection of a canonical page. This directive is intended to exclude the page from the index, not to manage the choice of a canonical page.

Also, checking one of the examples above, https://tracing.rs has `<meta name="robots" content="noindex">` in its source. Presumably the intent is that people should find the doc.rs page on search rather than the prerelease docs on https://tracing.rs. If the `tracing` crate set its documentation URL to https://tracing.rs (it doesn't), we would wind up in the tricky situation where _neither_ https://tracing.rs _nor_ https://docs.rs/tracing showed up in search indexes.

Maybe this isn't a big deal - I haven't surveyed the list of crates with documentation URLs that @pietroalbini thoughtfully provided to see how common it might be. But it makes me realize that adding `noindex` on docs.rs any time there is a documentation URL may be too aggressive.

**Comment by syphar** [comment]
so google can choose between the versioned pages on docs.rs?

**Comment by jsha** [comment]
An update from https://github.com/rust-lang/docs.rs/issues/1438#issuecomment-1209474723:

There are a fairly large number of pages that are not getting the `/latest/` treatment in Google's index because they have a documentation URL that points somewhere other than docs.rs, which means they don't have `<link rel="canonical">` (as I proposed above, and implemented). One particularly notable effect is that an _older_ version of a crate can have a non-docs.rs URL, which will mean the older version doesn't get `<link rel="canonical">`, and may itself get incorrectly selected as canonical.

I think the simplest solution is to apply `<link rel="canonical">` to all versions of all crates, and not make an exception for crates that have their own doc URL. Here's my reasoning:

For crates that have a self-hosted doc URL, we can't just point `rel="canonical"` at that doc URL. The doc URL could contain versioned URLs (like docs.rs); it could contain unversioned URLs (if only one version is hosted at a time); or it could be generic high-level documentation that doesn't match URLs one-for-one. Without know which, we'll get it wrong a good chunk of the time.

Given that, there's nothing we can do unilaterally to boost the ranking of that doc URL. Instead, we should make a mechanism available for crate authors to say "I would prefer my self-hosted documentation to show up on Google instead of docs.rs' documentation." For instance, we [could provide a package.metadata.docs.rs field](https://docs.rs/about/metadata), something like `noindex = true`. When that field is present on the _latest_ version of a crate, docs.rs would render _all_ versions of that crate with `<meta name="robots" content="noindex">`. Or we could use the `package.metadata.docs.rs.canonical-url` field [that @pietroalbini proposed in 2020](https://github.com/rust-lang/docs.rs/issues/74#issuecomment-682596113). Either way I think we need to treat the latest version of this metadata field as affecting all versions of the crate. And I think we should skip the banner.

Assuming folks here agree with that conclusion, we can uncouple the two issues: fixing canonicalization _within_ docs.rs, and offering a noindex option so crates can choose to boost their off-docs.rs documentation.

**Comment by jsha** [comment]
With #1792 released, instances of "Duplicate without user-selected canonical" have decreased, and nearly all of them are of the form `https://docs.rs/crate/cargo-bump/1.1.0`. In other words crate pages. We should give the same `<link rel="canonical">` treatment to crate pages, though it's not as crucial as for doc pages, since almost no-one goes to crate pages from search (4 out of 1000 top pages visited from search, according to Google Search Console).

**Comment by syphar** [comment]
This is awesome! thank you for driving this forward. 

My feeling would be that we should add the canonical url to these crate-pages too, for the sake of completenes, and then close this issue.

**Comment by jsha** [comment]
I've closed out #1438. For duplicates within docs.rs, we solved the problem by setting noindex on outdated versions.

I'm also closing out this issue. Based on the discussion above, automatically setting a `<link rel="canonical">` based on a crate's documentation field won't work, because we don't know what format the documentation is in and how to map individual pages to it. If we want to solve the original problem (docs.rs causes self-hosted docs to not appear in search), we'll do it by adding an explicit mechanism for crates to indicate noindex for their pages on docs.rs.

**Comment by andrewtj** [comment]
> If we want to solve the original problem (docs.rs causes self-hosted docs to not appear in search), we'll do it by adding an explicit mechanism for crates to indicate noindex for their pages on docs.rs.

Would including a snippet via rustdoc's `--html-in-header` feature work? 

```toml
[package.metadata.docs.rs]
rustdoc-args = ["--html-in-header", "noindex.html"]
```

noindex.html:
```html
<meta name="robots" content="noindex">
```
